book-content/chapters/12-the-weird-parts.md

```diff
@@ -0,0 +1,1653 @@
+We've now got a good understanding of most of TypeScript's features. Let's take it to the next level. By exploring some of the more unusual and lesser-known parts of TypeScript, we'll gain a deeper understanding of how it works.
+
+## The Evolving `any` Type
+
+While most of the time we want to have our types remain static, it is possible to create variables that can dynamically change their type like in JavaScript. This can be done with a technique called the "evolving `any`" which takes advantage of how variables are declared and inferred when no type is specified.
+
+To start, use `let` to declare the variable without a type, and TypeScript will infer it as `any`:
```

Comment:
it's actually not any, it's magic ;p

it's true that internally TS uses a special kind of `any` to track this:
```ts
var autoType = createIntrinsicType(TypeFlags.Any, "any", ObjectFlags.NonInferrableType, "auto");
```
and it's true that it can be (confusingly) displayed as `any` in tooltips but it's not `any`. If it would be `any` then the return type of this function should be `any`, right?
```ts
function test() {
  let foo;
  return foo;
}
```

But the return type here isn't any - it's `undefined` :) Matching what we can observe [here](https://www.typescriptlang.org/play/?ts=5.6.0-dev.20240721#code/GYVwdgxgLglg9mABFApgZygCgJSIN4BQiiANilIsHHANxGID0DxAegPz1Vz1OLsEBfIA)

The fact that `any` is sometimes displayed is because you check that a **write** positions - and we can write to it... well, anything. But checking such variables at read positions displays the correct/current type.

I might be really nit-picky here - especially given how it's displayed in tooltips. But how it's displayed in tooltip is a potential subject to change (there is an issue about it!). So in the future, the any-ness of it might not even be user-facing at all.